pkg/gpu/stats.go

```diff
@@ -0,0 +1,101 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package gpu
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/gpu/model"
+)
+
+// statsGenerator connects to the active stream handlers and generates stats for the GPU monitoring.
+type statsGenerator struct {
+	streamHandlers      map[model.StreamKey]*StreamHandler // streamHandlers contains the map of active stream handlers.
+	lastGenerationKTime int64                              // lastGenerationTime is the kernel time of the last stats generation.
+	currGenerationKTime int64                              // currGenerationTime is the kernel time of the current stats generation.
+	aggregators         map[uint32]*aggregator             // aggregators contains the map of aggregators
+	sysCtx              *systemContext                     // sysCtx is the system context with global GPU-system data
+}
+
+func newStatsGenerator(sysCtx *systemContext, currKTime int64, streamHandlers map[model.StreamKey]*StreamHandler) *statsGenerator {
+	return &statsGenerator{
+		streamHandlers:      streamHandlers,
+		aggregators:         make(map[uint32]*aggregator),
+		lastGenerationKTime: currKTime,
+		currGenerationKTime: currKTime,
+		sysCtx:              sysCtx,
+	}
+}
+
+func (g *statsGenerator) getStats(nowKtime int64) *model.GPUStats {
+	g.currGenerationKTime = nowKtime
+
+	for key, handler := range g.streamHandlers {
+		aggr := g.getOrCreateAggregator(key)
+		currData := handler.getCurrentData(uint64(nowKtime))
+		pastData := handler.getPastData(true)
+
+		if currData != nil {
+			aggr.processCurrentData(currData)
+		}
+
+		if pastData != nil {
+			aggr.processPastData(pastData)
+		}
+
+		if handler.processEnded {
+			aggr.processEnded = true
+		}
+	}
+
+	g.configureNormalizationFactor()
+
+	stats := model.GPUStats{
+		PIDStats: make(map[uint32]model.PIDStats),
+	}
+
+	for pid, aggregator := range g.aggregators {
+		stats.PIDStats[pid] = aggregator.getStats()
+	}
+
+	g.lastGenerationKTime = g.currGenerationKTime
+
+	return &stats
+}
+
+func (g *statsGenerator) getOrCreateAggregator(streamKey model.StreamKey) *aggregator {
+	aggKey := streamKey.Pid
+	if _, ok := g.aggregators[aggKey]; !ok {
+		g.aggregators[aggKey] = newAggregator(g.sysCtx)
+	}
+
+	g.aggregators[aggKey].lastCheckKtime = g.lastGenerationKTime
+	g.aggregators[aggKey].measuredIntervalNs = g.currGenerationKTime - g.lastGenerationKTime
+	return g.aggregators[aggKey]
```

Comment:
consider reversing the if clause, if the aggregator presents, just return it, otherwise create one and initialize the relevant fields.

that code would be much more readable + you don't need to execute lunes 75 and 76 that don't do anything when handling other streams of the same process

Reply:
In this case those fields aren't just initialized but refreshed, as the check times might have changed between runs

Reply:
so this logic is also stateful? hmm i feel it is unnecessary, but let me think on other options

Reply:
For the utilization statistic we always need the time that has passed since the last check, unfortunately there's no way around that I think. However, I don't really consider it stateful as the time is actually an input parameter to the aggregator.